{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "from scipy.stats import linregress\n",
    "import seaborn as sns\n",
    "# Model 3: Miniaturized 'real' CRISPEY Experiment\n",
    "\n",
    "#bottleneck_counts and extract_DNA are the same, copied for readability\n",
    "def bottleneck_counts(counts, bottleneck_size=800000):\n",
    "    #samples strains based from a poisson with lambda = their fraction in the library * bottleneck_size\n",
    "    #modeling the culture bottleneck when cells are diluted\n",
    "    lambdas = counts/counts.sum()*bottleneck_size\n",
    "    return np.array([np.random.poisson(i) for i in lambdas])\n",
    "\n",
    "def extract_DNA(counts, extraction_barcodes):\n",
    "    \n",
    "    '''\n",
    "    samples strains based on a poisson distribution with lambda = their fraction in the library * extraction_barcodes\n",
    "    modeling the culture bottleneck\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    counts : Numpy array (int)\n",
    "        Counts for strains at the end of a time point \n",
    "    extraction_barcodes : int\n",
    "        number of total barcodes to be extracted (average coverage in extraction * )\n",
    "   \n",
    "    Returns\n",
    "    -------\n",
    "    Numpy array\n",
    "       Array containing the final sequenced counts for each strain\n",
    "\n",
    "    '''\n",
    "    lambdas = counts/counts.sum()*extraction_barcodes\n",
    "    return np.array([np.random.poisson(i) for i in lambdas])\n",
    "\n",
    "def PCR_and_sequence(counts, num_reads, pcr_cycles, efficiency):\n",
    "    '''\n",
    "    simulates PCR process with some error (no clue what would be reasonable there)\n",
    "    then samples from the pcr counts as a poisson for each strain with\n",
    "    lambda = their fraction in the pcr counts * sequencing depth\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    counts : Numpy array (int)\n",
    "        Counts for strains after DNA extraction \n",
    "    num_reads : int\n",
    "        Rough number of sequencing reads\n",
    "    pcr_cycles: int\n",
    "        Number of PCR cycles\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    Numpy array\n",
    "       Array containing the final sequenced counts for each strain\n",
    "\n",
    "    '''\n",
    "    pcr = np.copy(counts)\n",
    "    for i in range(pcr_cycles):\n",
    "        num_amplified = np.random.binomial(pcr,efficiency)\n",
    "        pcr = (pcr-num_amplified)+num_amplified*2\n",
    "    sequence_lambdas = pcr/pcr.sum()*num_reads\n",
    "    return np.array([np.random.poisson(i) for i in sequence_lambdas])\n",
    "\n",
    "def calculate_fitness(s, e, num_gens):\n",
    "    #calculates fitness as a function of start and end counts\n",
    "    #currently not used\n",
    "    try:\n",
    "        fitness = 2**(np.log2((e-s)/s +1)/num_gens)\n",
    "        if fitness <.001:\n",
    "            fitness = None\n",
    "    except:\n",
    "        fitness = None\n",
    "    return fitness\n",
    "\n",
    "def simulate_crispey(fitness_size, bottleneck, \\\n",
    "                     num_strains = 400, num_tp = 6, gens_per_tp = 5, \\\n",
    "                     extraction_coverage = 4, sequencing_reads = 1000000, pcr_cycles = 24,\n",
    "                    num_reps = 1, pcr_efficiency = .9):\n",
    "    '''\n",
    "\n",
    "    Assuming a lot of poisson and other processes, simulates a Crispey competition\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    fitness_size : float\n",
    "        Scale for exponential distribution of relative fitness differences from 1 for strains in the competition. \n",
    "    bottleneck : int\n",
    "        Rough number of yeast passaged into the next timepoint, also used to derive the initial counts for strains\n",
    "    num_strains: int\n",
    "        Number of unique strains with fitness values being measured\n",
    "    num_tp: int\n",
    "        Number of time points in the competition when yeast are passaged/sampled.\n",
    "    gens_per_tp: int\n",
    "        Number of doublings a neutral yeast strain would go through in each timepoint\n",
    "    extraction_coverage: float\n",
    "        Average coverage of each strain in the extracted DNA (number of yeast genomes for each edit)\n",
    "    sequencing_reads: int\n",
    "        Rough number of reads to be sequenced\n",
    "    pcr_cycles: int\n",
    "        Number of PCR cycles performed on extracted DNA\n",
    "    num_reps: int\n",
    "        Number of replicate competitions performed\n",
    "    Returns\n",
    "    -------\n",
    "    output_df: Pandas DataFrame\n",
    "        DataFrame containing the sequenced counts of each strain in each time point for each replicate,\n",
    "        along with the true fitness value for each strain\n",
    "\n",
    "    See Also\n",
    "    --------\n",
    "    bottleneck_counts: helper function for the culturing bottleneck step\n",
    "    extract_DNA: helper function to perform DNA extraction step\n",
    "    PCR_and_sequence: helper function to perform PCR and sequencing steps\n",
    "    '''\n",
    "    #Having total starting counts equal to bottleneck size--realistic\n",
    "    starting_count_size = bottleneck/num_strains\n",
    "\n",
    "    #Sampling starting counts from exponential distribution to very uneven library distribution\n",
    "    starting_counts = np.round(np.random.exponential(starting_count_size, num_strains))\n",
    "    \n",
    "    #Sampling fitness effects from exponential distribution\n",
    "    fit_effects = np.random.exponential(fitness_size,num_strains)\n",
    "    \n",
    "    #Making fitness effects not all positive\n",
    "    selection_coefs = np.array([each*random.choice([-1,1]) for each in fit_effects])\n",
    "    fitness = 1+selection_coefs\n",
    "    output_df = pd.DataFrame({'True_fitness':fitness, 'T0_counts':starting_counts})\n",
    "    \n",
    "    #having all replicates start with the exact same starting counts\n",
    "    #moderately unrealistic\n",
    "    prev_counts = [starting_counts]*num_reps\n",
    "    for i in range(1, num_tp):\n",
    "        for j in range(num_reps):\n",
    "            #Simulating drift as a function of bottleneck size\n",
    "            drift = np.random.normal(0, 1/np.sqrt(bottleneck), num_strains)\n",
    "            \n",
    "            #simulating the growth of the strains in a timepoint\n",
    "            true_end_counts = prev_counts[j]*(2*fitness)**gens_per_tp+drift\n",
    "            true_end_counts = np.round(true_end_counts)\n",
    "            \n",
    "            extracted_counts = extract_DNA(true_end_counts, num_strains*extraction_coverage)\n",
    "            sequence_counts = PCR_and_sequence(extracted_counts, sequencing_reads, pcr_cycles, pcr_efficiency)\n",
    "            \n",
    "            output_df['T'+str(i)+'_rep'+str(j)+'_counts'] = sequence_counts\n",
    "            \n",
    "            prev_counts[j] = bottleneck_counts(true_end_counts,bottleneck)\n",
    "    output_df = output_df.fillna(0)\n",
    "    return output_df\n",
    "\n",
    "\n",
    "\n",
    "def significant_fitness_regression(crispey_df, replicate):\n",
    "    '''\n",
    "    Modifies a simulated Crispey dataset in place, adding regression slopes and p values for each strain\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    crispey_df : Pandas DataFrame\n",
    "        Simulated Crispey Counts Dataset \n",
    "    replicate : int\n",
    "        Replicate to calculate regressions on\n",
    "   \n",
    "    Returns\n",
    "    -------\n",
    "    Nothing\n",
    "    \n",
    "    '''\n",
    "    #modifies in place\n",
    "    c2 = crispey_df.filter(regex = 'rep'+str(replicate))\n",
    "    for column in c2.filter(regex = 'counts'):\n",
    "        c2[column] = np.log2((c2[column]+0.5)/c2[column].sum())\n",
    "    for i, row in c2.iterrows():\n",
    "        model_df = pd.DataFrame()\n",
    "        model_df['generations'] = range(0,25,5)\n",
    "        model_df['logfreq'] = row.filter(regex = 'counts').values.astype(float)\n",
    "\n",
    "        res = linregress(model_df['generations'],model_df['logfreq'])\n",
    "        crispey_df.loc[i,'regression_fitness_rep'+str(replicate)] = 1+res.slope\n",
    "        crispey_df.loc[i,'regression_pvalue_rep'+str(replicate)] = res.pvalue\n",
    "        \n",
    "def generate_fishers_table(hit_df, truefit,est_fit): \n",
    "    '''Generates a 2x2 contingency table for true fitness and fitness measured by simulation\n",
    "    Parameters\n",
    "    ----------\n",
    "    hit_df : Pandas DataFrame\n",
    "        Simulated Crispey hits dataframe with true and estimated fitness \n",
    "    truefit : str\n",
    "        column name for true fitness effects\n",
    "    est_fit : str\n",
    "        column name for estimated fitness effects\n",
    "   \n",
    "    Returns\n",
    "    -------\n",
    "    pandas DataFrame containing a 2x2 contingency table for the direction of fitness effects'''\n",
    "    \n",
    "    pospos = len(hit_df[(hit_df[truefit]>1)&(hit_df[est_fit]>1)])\n",
    "    posneg = len(hit_df[(hit_df[truefit]>1)&(hit_df[est_fit]<1)])\n",
    "    negpos = len(hit_df[(hit_df[truefit]<1)&(hit_df[est_fit]>1)])\n",
    "    negneg = len(hit_df[(hit_df[truefit]<1)&(hit_df[est_fit]<1)])\n",
    "    return pd.DataFrame({'index':['Negative Estimated Fitness', 'Positive Estimated Fitness'],'Negative True Fitness':[negneg, negpos], 'Positive True Fitness':[posneg,pospos]})\n",
    "\n",
    "def calculate_power(simdf, minimum_fitness, p = 0.05, truefit = 'True_fitness',\\\n",
    "                    est_fit= 'regression_fitness_rep0', est_p = 'regression_pvalue_rep0'):\n",
    "    '''Calculates power for a given replicate simulation\n",
    "    with regression to detect fitness effects a certain distance from 1\n",
    "    for a given p value threshold. (Also requires effect to be in the correct direction)\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    simdf : Pandas DataFrame\n",
    "        Simulated Crispey dataframe with true and estimated fitness \n",
    "    minimum_fitness: float\n",
    "        minimum value that true fitness effects must be away from 1 (neutral) to be considered\n",
    "        a hit for power analysis\n",
    "    p: float\n",
    "        p-value for an estimated fitness effect to be considered a hit\n",
    "    truefit : str\n",
    "        column name for true fitness effects\n",
    "    est_fit : str\n",
    "        column name for estimated fitness effects\n",
    "    est_p : str\n",
    "        column name for estimated fitness p values\n",
    "    Returns\n",
    "    -------\n",
    "    float : power as a decimal (0-1.0)\n",
    "    '''\n",
    "    #getting all \"real\" hits in the simulation\n",
    "    all_true_hits = simdf[abs(1-simdf[truefit])>minimum_fitness]\n",
    "    #getting all \"real\" hits in the simulation with a p value less than p \n",
    "    #from the fitness estimation and where the estimated fitness agrees with\n",
    "    #the direction of the true fitness\n",
    "    detected_true_hits = simdf[(abs(1-simdf[truefit])>minimum_fitness)&\\\n",
    "                               (simdf[est_p]<p)&(simdf[truefit]*simdf[est_fit]>0)]\n",
    "    power = len(detected_true_hits)/len(all_true_hits)\n",
    "    return power\n",
    "    \n",
    "def true_fitness_comp(simdf, rep, name_of_sim = ''):\n",
    "    '''plotting function for plotting scatter plot of measured fitness versus true fitness for\n",
    "    a CRISPEY simulation\n",
    "    \n",
    "    '''\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    scatter = ax.scatter(simdf['True_fitness'],simdf['regression_fitness_rep'+str(rep)],c = simdf['sigreg'+str(rep)])\n",
    "\n",
    "\n",
    "    plt.axvline(1, c = 'red', linestyle = 'dotted')\n",
    "    plt.axhline(1, c = 'red', linestyle = 'dotted')\n",
    "    plt.xlabel('True Fitness')\n",
    "    plt.ylabel('Estimated Fitness')\n",
    "    legend1 = ax.legend(*scatter.legend_elements(),\n",
    "                        loc=\"lower right\", title=\"p<.05\")\n",
    "    ax.add_artist(legend1)\n",
    "    plt.title(name_of_sim+' Rep '+str(rep+1)+' Simulation Regression Analysis'  )\n",
    "    \n",
    "def est_fitness_comp(simdf,  reps = [0,1], name_of_sim = '', filter_sig = []):\n",
    "    \n",
    "    '''plotting function for plotting scatter plot of measured fitness in different replicates\n",
    "    a CRISPEY simulation'''\n",
    "    #filter_sig can be rep number, None, or 'both'\n",
    "    if len(filter_sig) == 2:\n",
    "        for rep in filter_sig:\n",
    "            plotdf = simdf[(simdf['sigreg'+str(reps[0])])&(simdf['sigreg'+str(reps[1])])]\n",
    "    elif len(filter_sig):\n",
    "        plotdf = simdf[(simdf['sigreg'+str(filter_sig[0])])]\n",
    "    else:\n",
    "        plotdf = simdf.copy()\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    scatter = ax.scatter(plotdf['regression_fitness_rep'+str(reps[0])],plotdf['regression_fitness_rep'+str(reps[1])])\n",
    "\n",
    "\n",
    "    plt.axvline(1, c = 'red', linestyle = 'dotted')\n",
    "    plt.axhline(1, c = 'red', linestyle = 'dotted')\n",
    "    plt.xlabel('Rep1 Fitness')\n",
    "    plt.ylabel('Rep2 Fitness')\n",
    "    plt.title(name_of_sim+' Rep1 vs Rep2 Estimated Fitness')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Doing a parameter sweep for the extraction coverage (how many yeast have dna prepped),\n",
    "#and sequencing depth, and storing lots of diagnostics for each simulation\n",
    "#slow, next step would be optimizing the regression step or plugging into limma-voom or deseq\n",
    "\n",
    "sig_level = .05\n",
    "d = {'extraction coverage':[], 'total reads':[], 'replicate_correlation':[], \\\n",
    "     'rep1_true_fitness_corr':[],'rep2_true_fitness_corr':[],'replicate_corr_both_hits':[],\\\n",
    "    'rep1_true_fitness_corr_both_hits':[],'rep2_true_fitness_corr_both_hits':[],'number_of_hits':[],\n",
    "     'power_rep1':[], 'power_rep2':[]}\n",
    "for extraction_level in [100, 500,1000,2000,4000,5000, 10000,25000,100000]:\n",
    "    for sequencing_depth in [100,250,500,1000,2000,3000,5000,10000]:\n",
    "        for i in range(100):\n",
    "            sim = simulate_crispey(.005,800000, num_strains = 400, num_reps =2, \\\n",
    "                               extraction_coverage = extraction_level, sequencing_reads=sequencing_depth*400)\n",
    "            significant_fitness_regression(sim,0)\n",
    "            significant_fitness_regression(sim,1)\n",
    "            sim['sigreg0'] = sim['regression_pvalue_rep0']<sig_level\n",
    "            sim['sigreg1'] = sim['regression_pvalue_rep1']<sig_level\n",
    "            \n",
    "            both_hits = sim[(sim['sigreg0'])&(sim['sigreg1'])]\n",
    "            bcorr = both_hits.filter(regex = 'fitness').corr()\n",
    "            sim_fit_corr = sim.filter(regex = 'fitness').corr()\n",
    "            d['number_of_hits'].append(len(both_hits))\n",
    "            d['replicate_corr_both_hits'].append(bcorr.loc['regression_fitness_rep0','regression_fitness_rep1']**2)\n",
    "            d['rep1_true_fitness_corr_both_hits'].append(bcorr.loc['regression_fitness_rep0','True_fitness']**2)\n",
    "            d['rep2_true_fitness_corr_both_hits'].append(bcorr.loc['regression_fitness_rep1','True_fitness']**2)\n",
    "            d['extraction coverage'].append(extraction_level)\n",
    "            d['total reads'].append(sequencing_depth*400)\n",
    "            d['replicate_correlation'].append(sim_fit_corr.loc['regression_fitness_rep0','regression_fitness_rep1']**2)\n",
    "            d['rep1_true_fitness_corr'].append(sim_fit_corr.loc['regression_fitness_rep0','True_fitness']**2)\n",
    "            d['rep2_true_fitness_corr'].append(sim_fit_corr.loc['regression_fitness_rep1','True_fitness']**2)\n",
    "            d['power_rep1'].append(calculate_power(sim, .01, sig_level))\n",
    "            d['power_rep2'].append(calculate_power(sim, .01, sig_level,est_fit= 'regression_fitness_rep1', est_p = 'regression_pvalue_rep1'))\n",
    "        print('done with extraction level '+ str(extraction_level)+' and sequencing level '+str(sequencing_depth))\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "sim_results = pd.DataFrame(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s2 = sim_results.groupby(['extraction coverage', 'total reads']).agg('mean')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_results.index = sim_results['extraction coverage']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_results.drop('extraction coverage', inplace = True, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(s2.unstack()['rep2_true_fitness_corr'])\n",
    "plt.title('R^2 b/w True fitness and one Rep Measured Fitness for all simulated strains 100 sims each')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(s2.unstack()['rep1_true_fitness_corr_both_hits'], annot = s2.unstack()['number_of_hits'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(s2.unstack()['replicate_correlation'])\n",
    "plt.title('R^2 b/w replicates for all simulated strains 100 sims each')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(s2.unstack()['replicate_corr_both_hits'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CRISPEY_SIM",
   "language": "python",
   "name": "crispey_sim"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
